Adjust bmi160 driver to support both SPI and I2C

sjg20 opened the pull request:
This uses the SPI and I2C emulator frameworks to create a test that checks access to an accelerometer via both I2C and SPI.

At present the emulators are segmented by bus. Perhaps this should move to subsys/emul

sjg20 committed:
drivers: sensor: bmi160: Shorten the structs a bit

This driver uses verbose names for the config and data structures, which
makes it harder to see which one we are talking about. Shorten it, like
the bme280 driver.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: sensor: bmi160: Define accessors for the bmi160

Add inline functions to obtain the data and configuration.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: sensor: bmi160: Move SPI config to config

At present this driver only supports a single instance. It sets up some
of its config in the init routine. It is better to put config in
constant data so that multiple instances can be supported and RAM space
is minimised.

Update the driver accordingly.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: sensor: bmi160: Use only one read/write function

At present there are three separate read functions and two write
functions. This makes it harder to provide an interface that can work
with either SPI or I2C.

Use bmi160_read() for all reads and create a new bmi160_write()
function for all writes.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: sensor: bmi160: Rename 'spi' to 'bus'

This member holds the SPI bus pointer. Change its name to make that
more obvious and so that it can be used for an I2C bus also.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
emul: spi: bmi160: Prepare to support I2C

At present this driver only supports SPI. Refactor it so that SPI is
just one of the options. This does not change any functionality.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
emul: spi: emul_bmi160: Move logging to access functions

Move the logging for transactions to the functions that actually access
registers and sample data. This avoids needed to repeat this code when
I2C starts calling these functions.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
emul: spi: emul_bmi160: Rename 'spi' to 'bus'

Update the language to refer to a bus instead of SPI. This will make it
more applicable with I2C support is added.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
emul: spi: bmi160: Add support for I2C

At present this emulator only supports SPI. Add support for I2C as
well.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: sensor: bmi160: Prepare to support I2C

At present this driver only supports SPI. Refactor it so that SPI is
just one of the options. This does not change any functionality.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: sensor: bmi160: Add support for I2C

At present this driver only supports SPI. Add support for I2C as well,
following the model of BME280.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
tests: drivers: sensor: Test both I2C and SPI accelerometer

Now that we have a BMI160 driver and emulator pair that support both
I2C and SPI, add a test that runs over both buses.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
drivers: sensor: bmi160: Use 'data' for the data pointer

Currently a 'bmi160' pointer is used to point to the driver data. This
confusing, as the driver uses both data and config. Rename the variable
to 'data' like the bme280 driver.

Signed-off-by: Simon Glass <sjg@chromium.org>

sjg20 committed:
doc: emulator: Add basic documentation

Add some basic documentation for Zephyr emulators. This is expanded
from https://github.com/zephyrproject-rtos/zephyr/issues/27531 and
recent commits.

Signed-off-by: Simon Glass <sjg@chromium.org>

jettr commented on tests/drivers/sensor/accel/src/main.c:
```diff
@@ -45,11 +46,25 @@ void test_sensor_accel_basic(void)
+{
+	run_tests_on_accel(DT_LABEL(DT_ALIAS(accel_0)));
+
+#ifdef DT_N_ALIAS_accel_1
```
```suggestion
#ifdef DT_ALIAS(accel_1)
```
Use the devicetree.h macros to generate this name for you, right?

Couldn't you even use `IS_ENABLED(DT_ALIAS(accel1))`?

sjg20 replied:
I can't use #ifdef on an expression. This is not a CONFIG so IS_ENABLED() doesn't work. I am basically checking for the presence of an alias in the devicetree_unfixed.h file:

/* Existence and alternate IDs: */
#define DT_N_S_i2c_100_S_bmi_68_EXISTS 1
#define DT_N_ALIAS_accel_1       DT_N_S_i2c_100_S_bmi_68
#define DT_N_INST_0_bosch_bmi160 DT_N_S_i2c_100_S_bmi_68
#define DT_N_NODELABEL_bmi_i2c   DT_N_S_i2c_100_S_bmi_68

jettr commented on subsys/emul/spi/emul_bmi160.c:
```diff
@@ -2,6 +2,9 @@
  * SPDX-License-Identifier: Apache-2.0
+ *
+ * Emulator for the Boche BMI160 accelerometer / gyro. This supports basic
+ * init and reading of canned samples. It supports both I2C and SPI buses.
```
The files location doesn't make sense anymore if this file support both SPI and I2C.  Can you put this in subsys/emul/sensors/emul_bmi160.c instead?

sjg20 replied:
OK done

jettr commented on drivers/sensor/bmi160/bmi160.h:
```diff
@@ -398,8 +398,29 @@ struct bmi160_range {
 
+#define BMI160_BUS_SPI		DT_ANY_INST_ON_BUS_STATUS_OKAY(spi)
+
+union bmi160_bus_cfg {
```
This is personal preference, but can you nest this union down a level, please?

```
struct bmi_bus_cfg {
   union {
#if BM160_BUS_SPI
	const struct spi_config *spi_cfg;
#endif
  }
};
```

It will take the same about of space and it will give you a place to put common config if the need arise (and you don't have to pass around a dirty union in your function signatures)

sjg20 replied:
done

jettr reviewed (changes requested):
Overall looks good. Just a few minor comments

sjg20 committed:
emul: spi: bmi160: Move to top-level directory

This emulator currently only supports SPI. Before making it also
support I2C, move it up a directory to avoid I2C uses missing it.

Signed-off-by: Simon Glass <sjg@chromium.org>
